beacon-chain/utils/shuffle.go

```diff
@@ -3,37 +3,60 @@ package utils
 
 import (
 	"errors"
+	"math"
 
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/prysmaticlabs/prysm/shared/hashutil"
 	"github.com/prysmaticlabs/prysm/shared/params"
 )
 
 // ShuffleIndices returns a list of pseudorandomly sampled
-// indices. This is used to use to select attesters and proposers.
-func ShuffleIndices(seed common.Hash, validatorList []uint32) ([]uint32, error) {
-	// Since we are consuming 3 bytes of entropy at a time in the loop,
+// indices. This is used to shuffle validators on ETH2.0 beacon chain.
+func ShuffleIndices(seed common.Hash, indicesList []uint32) ([]uint32, error) {
+	// Each entropy is consumed from the seed in randBytes chunks.
+	randBytes := int(params.BeaconConfig().RandBytes)
+	upperBound := uint64(math.Pow(2, float64(randBytes) * 8)) - 1
+
+	// Since we are consuming randBytes of entropy at a time in the loop,
 	// we have a bias at 2**24, this check defines our max list size and is used to remove the bias.
 	// more info on modulo bias: https://stackoverflow.com/questions/10984974/why-do-people-say-there-is-modulo-bias-when-using-a-random-number-generator.
-	if uint64(len(validatorList)) >= params.BeaconConfig().ModuloBias {
-		return nil, errors.New("exceeded upper bound for validator shuffle")
+	if uint64(len(indicesList)) >= upperBound {
+		return nil, errors.New("input list exceeded upper bound and reached modulo bias")
 	}
 
+	// Rehash the seed to obtain a new pattern of bytes.
 	hashSeed := hashutil.Hash(seed[:])
-	validatorCount := len(validatorList)
-
-	// Shuffle stops at the second to last index.
-	for i := 0; i < validatorCount-1; i++ {
-		// Convert every 3 bytes to random number, replace validator index with that number.
-		for j := 0; j+3 < len(hashSeed); j += 3 {
-			swapNum := int(hashSeed[j] + hashSeed[j+1] + hashSeed[j+2])
-			remaining := validatorCount - i
-			swapPos := swapNum%remaining + i
-			validatorList[i], validatorList[swapPos] = validatorList[swapPos], validatorList[i]
+	totalCount := len(indicesList)
+	index := 0
+	for index < totalCount-1 {
+		// Iterate through the hashSeed bytes in chunks of size randBytes.
+		for i := 0; i < 32-(32%randBytes); i += randBytes {
+			// Determine the number of indices remaining and exit if last index reached.
+			remaining := uint64(totalCount - index)
+			if remaining == 1 {
+				break
+			}
+			// Read randBytes of hashSeed as a 3 x randBytes big-endian integer.
```

Comment:
Do we know why this is 3? Where did this number come from?

Reply:
Is it 3 because randBytes is 3 or it’s some magic number that there can be a maximum of 2^24 validators? Is there some assertion that there can never be more than that number? Btw it’s only ~16.77MM validators. 

I think this needs more context here. I don’t understand it and I think other readers won’t understand it either. What do you think?

Reply:
My mistake, it should be 8 x randBytes and I'll update the comment 

It's assuming max validator account is 2^24 given the total eth supply

Reply:
Did you update the comment? It still says `3 x randBytes`